sql/src/test/java/org/apache/druid/sql/calcite/CalciteIngestionDmlTest.java

```diff
@@ -0,0 +1,339 @@
+import org.apache.druid.sql.SqlLifecycleFactory;
+import org.apache.druid.sql.calcite.external.ExternalDataSource;
+import org.apache.druid.sql.calcite.parser.DruidSqlInsert;
+import org.apache.druid.sql.calcite.planner.Calcites;
+import org.apache.druid.sql.calcite.planner.PlannerConfig;
+import org.apache.druid.sql.calcite.planner.PlannerContext;
+import org.apache.druid.sql.calcite.util.CalciteTests;
+import org.hamcrest.CoreMatchers;
+import org.hamcrest.Matcher;
+import org.hamcrest.MatcherAssert;
+import org.junit.After;
+import org.junit.Assert;
+import org.junit.internal.matchers.ThrowableMessageMatcher;
+
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+
+public class CalciteIngestionDmlTest extends BaseCalciteQueryTest
+{
+  protected static final Map<String, Object> DEFAULT_CONTEXT =
+      ImmutableMap.<String, Object>builder()
+                  .put(PlannerContext.CTX_SQL_QUERY_ID, DUMMY_SQL_ID)
+                  .build();
+
+  protected static final RowSignature FOO_TABLE_SIGNATURE =
+      RowSignature.builder()
+                  .addTimeColumn()
+                  .add("cnt", ColumnType.LONG)
+                  .add("dim1", ColumnType.STRING)
+                  .add("dim2", ColumnType.STRING)
+                  .add("dim3", ColumnType.STRING)
+                  .add("m1", ColumnType.FLOAT)
+                  .add("m2", ColumnType.DOUBLE)
+                  .add("unique_dim1", HyperUniquesAggregatorFactory.TYPE)
+                  .build();
+
+  protected final ExternalDataSource externalDataSource = new ExternalDataSource(
+      new InlineInputSource("a,b,1\nc,d,2\n"),
+      new CsvInputFormat(ImmutableList.of("x", "y", "z"), null, false, false, 0),
+      RowSignature.builder()
+                  .add("x", ColumnType.STRING)
+                  .add("y", ColumnType.STRING)
+                  .add("z", ColumnType.LONG)
+                  .build()
+  );
+
+  protected boolean didTest = false;
+
+  @After
+  @Override
+  public void tearDown() throws Exception
+  {
+    super.tearDown();
+
+    // Catch situations where tests forgot to call "verify" on their tester.
+    if (!didTest) {
+      throw new ISE("Test was not run; did you call verify() on a tester?");
+    }
+  }
+
+  protected String externSql(final ExternalDataSource externalDataSource)
+  {
+    try {
+      return StringUtils.format(
+          "TABLE(extern(%s, %s, %s))",
+          Calcites.escapeStringLiteral(queryJsonMapper.writeValueAsString(externalDataSource.getInputSource())),
+          Calcites.escapeStringLiteral(queryJsonMapper.writeValueAsString(externalDataSource.getInputFormat())),
+          Calcites.escapeStringLiteral(queryJsonMapper.writeValueAsString(externalDataSource.getSignature()))
+      );
+    }
+    catch (JsonProcessingException e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+  protected Map<String, Object> queryContextWithGranularity(Granularity granularity)
```

Comment:
Can we extract this as a static method in the common class for the insert tests? (Similar comment as the above)